test/tests/selftest/pbs_requirements_decorator.py

```diff
@@ -0,0 +1,84 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2018 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.selftest import *
+
+
+@tags('smoke')
+@requirements(num_servers=2, num_comms=2)
+class TestRequirementsDecorator(TestSelf):
+
+    """
+    This test suite is to verify functionality of requirements
+    decorator. This test suite when run on single node without any
+    hostnames passed in -p or param file, tests run and skip functionality
+    of this decorator
+
+    """
+    @requirements(num_servers=1, num_comms=1)
```

Comment:
I feel like these tests aren't really helpful. One will just see "skipped" in the output instead of "failed" or "passed", and it'll be difficult to know whether "skipped" means that the test passed as expected, or if it failed.

Not sure if this approach will work for you as the decorator doesn't seem to be doing the actual check itself (you seem to be doing the actual check inside startTest()), but do check this out and see if you can write tests similarly:
https://stackoverflow.com/questions/34648337/how-to-test-a-decorator-in-a-python-package

Reply:
With the current approach mentioned in my previous note about check not being done in requirements decorator; the most I could do is verify the requirements at test case level by extracting the REQUIREMENTS_KEY and verify that it is set. I have modified test_tc_run with this check. The execution of test_tc_run indicates test case level requirements are effective over the test suite level requirements and rest of the tests getting skipped indicate working skip functionality. 
I have kept the rest part of tests as is. Please let me know your suggestions.